### PR TITLE
Soft deadlines

### PR DESCRIPTION
**Pull request by FeisEater**

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -95,6 +95,7 @@
     t.string   "valgrind_strategy"
     t.boolean  "code_review_requests_enabled",     default: false, null: false
     t.boolean  "run_tests_locally_action_enabled", default: true,  null: false
+    t.text     "soft_deadline_spec"
```
Comment (jamox) [inline db/schema.rb]: Wouldn't string suffice?

**Comment by jamox** [comment]
Maybe we should also be updating the [usermanual](https://github.com/testmycode/tmc-server/blob/master/spec/usermanual/instructors.html.erb)

File: db/schema.rb (hunk 1 above)
Comment (jruo) [inline db/schema.rb]: Probably yes, but the other _specs are text too. Should I change them?

**Comment by jruo** [comment]
We are updating the manual right now. It will be in its own branch.

**Comment by jruo** [comment]
Arto thought that students don't need to know about hard deadlines, but I changed it that it now shows hard deadline if soft deadline is not specified

File: db/schema.rb (hunk 1 above)
Comment (jamox) [inline db/schema.rb]: ok, we can leave it as it is.

**Comment by jamox** [comment]
One question remains, what happens to current courses without soft deadline spec?

**Comment by jruo** [comment]
They should work exaclty the same. Soft deadline specs can be null, and they don't change the way the current system works.

File: app/models/exercise_group.rb [hunk 2]
```diff
@@ -75,4 +80,17 @@ def group_unlock_conditions=(unlock_conditions)
...
+  private
+
+  def group_deadline(method)
+    exercises(false).map { |n| n.send(method) }.first
+  end
+
+  def set_group_deadline(method, deadline)
+    exercises(false).each do |e|
+      e.send(method, deadline)
+      e.save
```
Comment (jamox) [inline app/models/exercise_group.rb]: save!

File: spec/models/exercise_spec.rb [hunk 3]
```diff
@@ -99,6 +99,14 @@ def set_deadline(ex, t)
     end
   end
 
+  def set_soft_deadline(ex, t)
+    if t.is_a? Array
+      ex.soft_deadline_spec = t.to_json
+    else
+      ex.soft_deadline_spec = [t.to_s].to_json
+    end
+  end
```
Comment (jamox) [inline spec/models/exercise_spec.rb]: I don't like the .is_a? -- could we maybe survive with flatten?
And what t is if it is not an array

**Comment by jruo** [comment]
Updated

File: app/controllers/courses_controller.rb [hunk 4]
```diff
@@ -105,8 +105,10 @@ def save_deadlines
 
     groups = group_params
     groups.each do |name, deadlines|
-      json_array = [deadlines[:static], deadlines[:unlock]].to_json
-      @course.exercise_group_by_name(name).group_deadline=(json_array)
+      soft_deadlines = [deadlines[:soft][:static], deadlines[:soft][:unlock]].to_json
+      hard_deadlines = [deadlines[:hard][:static], deadlines[:hard][:unlock]].to_json
```
Comment (jamox) [inline app/controllers/courses_controller.rb]: should we somehow validate the input/ is it to be validated at some point?

Comment (jruo) [inline app/controllers/courses_controller.rb]: When we pass the input to the exercises, they try to parse it into a spec. If it fails, they raise an invalidsyntaxerror which is caught at the end of this method.

**Pull request merged**